### PR TITLE
39 vscode extension better status bar information

### LINKED ISSUE #39: vscode extension: better status bar information
Right now, when verification fails, the status bar shows "Verified". This is very confusing. Ideally, we'd distinguish between verification errors, type errors, syntax errors, timeouts, and successful verification.

In addition, the extension seems to show "Et tu Brute?" in the status bar if the verifier never started. This is the case if the server does not automatically start (which is unfortunately the default right now). If the server never started, we should print something like "Caesar not started".

### PR DESCRIPTION
**Pull request by umutdural**
- The texts on the status bar are replaced to better reflect the current state of Caesar.
- Command links in status bar tooltip are now working. 
- Tooltip changes (e.g. visible command links) based on the state of Caesar.
- LSP server doesn't delete the information related to all files upon receiving a verify request. Instead it only deletes the information about the file that the verify request is sent for.
- Multiple visible tabs can be decorated w.r.t. the verification statuses of each file. Status bar tooltip displays information for each visible file in a vertical list.

Closes #39.

### CODE REVIEW

File: vscode-ext/src/StatusBarComponent.ts [hunk 1]
```diff
@@ -38,36 +47,138 @@ export class StatusBarComponent {
...
             }
+
+            this.renderTooltip();
             this.view.show();
         } else {
             this.view.hide();
         }
     }
+
+    private extendedTooltip(extension: vscode.MarkdownString) {
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:107]: Seems like there is a return type missing? I wonder why TypeScript doesn't complain about this.

Also, I think this method should be inlined. As far as I can tell, it's just used once.

File: vscode-ext/src/StatusBarComponent.ts [hunk 2]
```diff
@@ -38,36 +47,138 @@ export class StatusBarComponent {
...
+        if (this.tooltipExtension.value === "") {
+            this.view.tooltip = this.tooltipBase;
+            this.view.tooltip.isTrusted = true;
+            return;
+        }
+        this.view.tooltip = this.extendedTooltip(this.tooltipExtension);
+        this.view.tooltip.isTrusted = true;
+    }
+
+    private countResults(results: [Range, VerifyResult][]): [number, number, number] {
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:121]: I don't see a reason why this is a member function. Either make it a freestanding function or just inline its definition as well (I prefer the latter).

File: vscode-ext/src/StatusBarComponent.ts [hunk 3]
```diff
@@ -1,15 +1,29 @@
...
+
 export class StatusBarComponent {
 
     private enabled: boolean;
-    private status: ServerStatus = ServerStatus.Stopped;
+    private verifyStatus: DocumentMap<[Range, VerifyResult][]> = new DocumentMap();
+    private serverStatus: ServerStatus = ServerStatus.NotStarted;
     private view: StatusBarItem;
+    private tooltipBase: vscode.MarkdownString = stoppedTooltipBase;
+    private tooltipExtension: vscode.MarkdownString = new vscode.MarkdownString();
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:26]: Maybe rename this to `tooltipStatusText`? The current name is a bit confusing to me.

File: vscode-ext/src/StatusBarComponent.ts [hunk 4]
```diff
@@ -1,15 +1,29 @@
...
+        "[Start Caesar](command:caesar.startServer)", true);
+
 export class StatusBarComponent {
 
     private enabled: boolean;
-    private status: ServerStatus = ServerStatus.Stopped;
+    private verifyStatus: DocumentMap<[Range, VerifyResult][]> = new DocumentMap();
+    private serverStatus: ServerStatus = ServerStatus.NotStarted;
     private view: StatusBarItem;
+    private tooltipBase: vscode.MarkdownString = stoppedTooltipBase;
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:25]: Maybe rename this to `tooltipMenuText`?

File: vscode-ext/src/StatusBarComponent.ts [hunk 5]
```diff
@@ -38,36 +47,138 @@ export class StatusBarComponent {
 
         // listen to verifier updates
         verifier.client.onStatusUpdate((status) => {
-            this.status = status;
+            this.serverStatus = status;
+            this.render();
+        });
+
+        verifier.client.onVerifyResult((document, results) => {
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:54]: If a document is closed, is it removed from the internal `verifyStatus` map?

**Review by Philipp15b — CHANGES_REQUESTED** [review]
Thanks a lot! This will eliminate a lot of confusion I've seen with people using Caesar. I've left some minor comments.

There's also an issue I've discovered: If there's a syntax error, a name cannot be resolved, or if there's a type error, then the status bar will falsely show "Verified!". I don't know where in the code this should be fixed, but it's definitely very confusing.

File: src/ast/diagnostic.rs [hunk 6]
```diff
@@ -157,6 +157,13 @@ impl Files {
         self.files.iter().find(|file| &file.path == path)
     }
 
+    pub fn find_uri(&self, document_id: TextDocumentIdentifier) -> Option<&Arc<StoredFile>> {
```
Comment (Philipp15b) [inline src/ast/diagnostic.rs:160]: Is this method necessary? Why not just call `find(SourceFilePath::Lsp(ident))`?

File: src/servers/lsp.rs [hunk 7]
```diff
@@ -195,9 +196,22 @@ impl LspServer {
...
+                let params: DidCloseTextDocumentParams =
                     notification.extract("textDocument/didClose")?;
-                // TODO: remove file?
+
+                let file_id = self
+                    .files
+                    .lock()
+                    .unwrap()
+                    .find_uri(params.text_document.clone())
+                    .unwrap_or_else(|| {
```
Comment (Philipp15b) [inline src/servers/lsp.rs:207]: Possibly just turn this into an `.expect("Could not find file id for document")`. The `unwrap_or_else` with a panic is a bit weird to me.

File: vscode-ext/src/StatusBarComponent.ts [hunk 8]
```diff
@@ -38,36 +48,156 @@ export class StatusBarComponent {
...
+
+
+    private handleReadyStatus() {
+        let tooltipString = new vscode.MarkdownString("", true);
+
+        let someError = false;
+        let someVerified = false;
+
+        const editorsWithResults = vscode.window.visibleTextEditors.filter((editor) => {
+            return editor.document.languageId === "heyvl" // Only HeyVL files
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:134]: Why do you need to check for HeyVL files? Shouldn't `this.verifyStatus` already only contain just HeyVL files?

File: vscode-ext/src/StatusBarComponent.ts [hunk 9]
```diff
@@ -38,36 +48,156 @@ export class StatusBarComponent {
...
+        if (editorsWithResults.length === 0) {
+            // No HeyVL files open or no results yet
+            this.view.text = "$(thumbsup-filled) Caesar Ready";
+            return;
+        }
+
+        for (const editor of editorsWithResults) {
+            const document_id: TextDocumentIdentifier = { uri: editor.document.uri.toString() };
+
+            // The default will not be used because the filter above ensures that the document is in the map.
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:147]: Why handle the default case then with `?? []`? Why not just throw an error then?

File: vscode-ext/src/StatusBarComponent.ts [hunk 10]
```diff
@@ -38,36 +48,156 @@ export class StatusBarComponent {
...
             }
+
+            this.renderTooltip();
             this.view.show();
         } else {
             this.view.hide();
         }
     }
+
+    private renderTooltip() {
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:116]: Maybe add a comment that this renders the tooltip with the text provided in `this.tooltipMenuText`? Alternatively, just add `tooltipMenuText` as a parameter to this function and don't have it as an attribute at all?

File: vscode-ext/src/StatusBarComponent.ts [hunk 11]
```diff
@@ -38,36 +48,156 @@ export class StatusBarComponent {
...
+            this.view.tooltip = this.tooltipMenuText;
+            this.view.tooltip.isTrusted = true;
+            return;
+        }
+        this.view.tooltip = new vscode.MarkdownString(this.tooltipMenuText.value + "\n\n --- \n" + this.tooltipStatusText.value, true);
+        this.view.tooltip.isTrusted = true;
+    }
+
+
+    private handleReadyStatus() {
```
Comment (Philipp15b) [inline vscode-ext/src/StatusBarComponent.ts:127]: Maybe call this `renderStatusReady`

**Review by Philipp15b — CHANGES_REQUESTED** [review]
Thanks a lot! I've left a few very minor comments. After those we can merge!

File: src/servers/lsp.rs (hunk 7 above)
Comment (umutdural) [reply src/servers/lsp.rs:207]: Clippy complains about the `expect` and suggests the `unwrap_or_else` version. I think this is why the `unwrap_or_else` version is previously preferred in the `run_server` method as well.

**Review by Philipp15b — APPROVED** [review]
Thanks a lot!

**Pull request merged**